controllers/api/auth_test.go

```diff
@@ -0,0 +1,283 @@
+	"go-google-scraper-challenge/initializers"
+	"net/http"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+type LoginResponse struct {
+	AccessToken  string `json:"access_token"`
+	ExpiresIn    int    `json:"expires_in"`
+	RefreshToken string `json:"refresh_token"`
+	TokenType    string `json:"token_type"`
+}
+
+type ErrorResponse struct {
+	Error            string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+}
+
+var _ = Describe("AuthController", func() {
+	Describe("POST /login", func() {
+		Context("given valid params", func() {
+			It("responses with status ok", func() {
+				FabricateUser("dev@nimblehq.co", "password")
+				oauth_client := FabricateOAuthClient()
+				body := RequestBody(map[string]string{
+					"client_id":     oauth_client.ClientID,
+					"client_secret": oauth_client.ClientSecret,
+					"username":      "dev@nimblehq.co",
+					"password":      "password",
+					"grant_type":    "password",
+				})
+				response := MakeRequest("POST", "/api/v1/login", body)
+
+				Expect(response.StatusCode).To(Equal(http.StatusOK))
+			})
+
+			It("response with token data", func() {
+				FabricateUser("dev@nimblehq.co", "password")
+				oauth_client := FabricateOAuthClient()
+				body := RequestBody(map[string]string{
+					"client_id":     oauth_client.ClientID,
+					"client_secret": oauth_client.ClientSecret,
+					"username":      "dev@nimblehq.co",
+					"password":      "password",
+					"grant_type":    "password",
+				})
+				response := MakeRequest("POST", "/api/v1/login", body)
+				responseBody := LoginResponse{}
+				GetJSONResponseBody(response, &responseBody)
+
+				Expect(responseBody.AccessToken).NotTo(BeNil())
+				Expect(responseBody.RefreshToken).NotTo(BeNil())
+				Expect(responseBody.ExpiresIn).NotTo(BeNil())
+				Expect(responseBody.TokenType).NotTo(BeNil())
+			})
+		})
+
+		Context("given missing params", func() {
+			Context("given NO client id", func() {
+				It("response with invalid request error", func() {
+					FabricateUser("dev@nimblehq.co", "password")
+					oauth_client := FabricateOAuthClient()
+					body := RequestBody(map[string]string{
+						"client_id":     "",
+						"client_secret": oauth_client.ClientSecret,
+						"username":      "dev@nimblehq.co",
+						"password":      "password",
+						"grant_type":    "password",
+					})
+					response := MakeRequest("POST", "/api/v1/login", body)
+					responseBody := ErrorResponse{}
+					GetJSONResponseBody(response, &responseBody)
+
+					Expect(response.StatusCode).To(Equal(http.StatusBadRequest))
+					Expect(responseBody.Error).To(Equal("Bad Request"))
+					Expect(responseBody.ErrorDescription).To(Equal("The request is missing a required parameter, includes an invalid parameter value, includes a parameter more than once, or is otherwise malformed"))
+				})
+			})
+
+			Context("given NO client secret", func() {
+				It("response with invalid request error", func() {
```

Comment:
Maybe just bad request would be clear enough as it's the actually HTTP error:

```suggestion
				It("returns a bad request error", func() {
```

Reply:
@olivierobert fixed on [79969f4](https://github.com/carryall/go-google-scraper-challenge/pull/26/commits/79969f4341e3cdb551b3073a17c27dfd4d19a2bd)